spec/index.bs

```diff
@@ -2428,114 +2428,115 @@ A {{GPUBindGroupLayout}} object has the following internal slots:
+        **Called on:** {{GPUDevice}} |this|.
 
-        <dfn>storage buffer validation</dfn>: There must be {{GPULimits/maxStorageBuffersPerShaderStage|GPULimits.maxStorageBuffersPerShaderStage}} or
-            fewer |bindingDescriptor|s of type {{GPUBindingType/storage-buffer}} visible on each shader stage in |descriptor|.
+        **Arguments:**
+        <pre class=argumentdef for="GPUDevice/createBindGroupLayout(descriptor)">
+            |descriptor|:
+        </pre>
+
+        **Returns:** {{GPUBindGroupLayout}}
+
+        1. If any of the following conditions are unsatisfied:
+            <div class=validusage>
+                - |this| is a [=valid=] {{GPUDevice}}.
+                - Each |bindingDescriptor|.{{GPUBindGroupLayoutEntry/binding}} in |descriptor| is unique.
+                - There is {{GPULimits/maxUniformBuffersPerShaderStage|GPULimits.maxUniformBuffersPerShaderStage}}
+                    or fewer |bindingDescriptor|s of type {{GPUBindingType/uniform-buffer}} visible
+                    on each shader stage in |descriptor|.
+                - There is {{GPULimits/maxDynamicUniformBuffersPerPipelineLayout|GPULimits.maxDynamicUniformBuffersPerPipelineLayout}}
+                    or fewer |bindingDescriptor|s of type {{GPUBindingType/uniform-buffer}} with
+                    {{GPUBindGroupLayoutEntry/hasDynamicOffset}} set to `true` visible on any shader
+                    stage in |descriptor|.
+                -  There is {{GPULimits/maxStorageBuffersPerShaderStage|GPULimits.maxStorageBuffersPerShaderStage}}
+                    or fewer |bindingDescriptor|s of type {{GPUBindingType/storage-buffer}} visible
+                    on each shader stage in |descriptor|.
+                - There is {{GPULimits/maxDynamicStorageBuffersPerPipelineLayout|GPULimits.maxDynamicStorageBuffersPerPipelineLayout}}
+                    or fewer |bindingDescriptor|s of type {{GPUBindingType/storage-buffer}} with
+                    {{GPUBindGroupLayoutEntry/hasDynamicOffset}} set to `true` visible on any shader
+                    stage in |descriptor|.
+                - There is {{GPULimits/maxSampledTexturesPerShaderStage|GPULimits.maxSampledTexturesPerShaderStage}}
+                    or fewer |bindingDescriptor|s of type {{GPUBindingType/sampled-texture}} visible
+                    on each shader stage in |descriptor|.
+                - There is {{GPULimits/maxStorageTexturesPerShaderStage|GPULimits.maxStorageTexturesPerShaderStage}}
+                    or fewer |bindingDescriptor|s of type {{GPUBindingType/readonly-storage-texture}}
+                    and {{GPUBindingType/writeonly-storage-texture}} visible on each shader stage in
+                    |descriptor|.
+                - There is {{GPULimits/maxSamplersPerShaderStage|GPULimits.maxSamplersPerShaderStage}}
+                    or fewer |bindingDescriptor|s of type {{GPUBindingType/sampler}} visible on each
+                    shader stage in |descriptor|.
```

Comment:
I don't think there's a strong reason to have two Valid Usage boxes (normally we only need that if one is on a different timeline). How about having an extra "for each" inside here and moving all those validation steps here?

Reply:
Holy freaking validation block, Batman! 😆 Took a crack at it but I'm not thrilled with how "noisy" the rendered version looks. Happy to iterate on it more if you feel like it needs it.

Reply:
ugh, yeah, it is huge. And really squished too because the column has gotten narrow at this point.... Maybe this is a place where we should un-inline a "validate GPUBindGroupLayoutEntry" function?